tests/versioned/koa.tap.js

```diff
@@ -1,52 +1,162 @@
 
-  t.beforeEach(function(done) {
+  t.beforeEach((done) => {
     helper = utils.TestAgent.makeInstrumented()
     helper.registerInstrumentation({
       moduleName: 'koa',
       type: 'web-framework',
       onRequire: require('../../lib/instrumentation')
     })
-    var Koa = require('koa')
+    const Koa = require('koa')
     app = new Koa()
     done()
   })
 
-  t.afterEach(function(done) {
+  t.afterEach((done) => {
     server && server.close()
     app = null
     helper && helper.unload()
     done()
   })
 
-  t.test('produces transaction trace with multiple middleware', function(t) {
+  t.test('names the transaction after the middleware that sets the body', (t) => {
+    app.use(function one(ctx, next) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('one-start')
+      return next().then(() => tx.nameState.appendPath('one-end'))
+    })
+
+    app.use(function two(ctx) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('two')
+      ctx.body = 'done'
+    })
+
+    helper.agent.on('transactionFinished', (tx) => {
+      t.equal(
+        tx.name,
+        'WebTransaction/WebFrameworkUri/Koa/GET//one-start/two',
+        'should have name without post-response name info'
+      )
+      t.end()
+    })
+
+    run()
+  })
+
+  t.test('names the transaction after the last middleware that sets the body', (t) => {
+    app.use(function one(ctx, next) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('one-start')
+      return next().then(() => tx.nameState.appendPath('one-end'))
+    })
+
+    app.use(function two(ctx, next) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('two')
+      ctx.body = 'not actually done'
+      return next()
+    })
+
+    app.use(function three(ctx) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('three')
+      ctx.body = 'done'
+    })
+
+    helper.agent.on('transactionFinished', (tx) => {
+      t.equal(
+        tx.name,
+        'WebTransaction/WebFrameworkUri/Koa/GET//one-start/two/three',
+        'should have name without post-response name info'
+      )
+      t.end()
+    })
+
+    run()
+  })
+
+  t.test('names the transaction off the status setting middleware', (t) => {
+    app.use(function one(ctx, next) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('one-start')
+      return next().then(() => tx.nameState.appendPath('one-end'))
+    })
+
+    app.use(function two(ctx) {
+      const tx = helper.getTransaction()
+      tx.nameState.appendPath('two')
+      ctx.status = 200
+    })
+
+    helper.agent.on('transactionFinished', (tx) => {
+      t.equal(
+        tx.name,
+        'WebTransaction/WebFrameworkUri/Koa/GET//one-start/two',
+        'should have name without post-response name info'
+      )
+      t.end()
+    })
+
+    run()
+  })
+
+  t.test('names the transaction when body set even if status set after', (t) => {
```

Comment:
I'm curious why this is desirable / why it would make sense to a consumer? My lack of familiarity w/ Koa likely showing here.

Reply:
My thought is that the middleware that sets the body is more likely considered the endpoint than one that sets the status code. Take this example sequence:

```js
app.use(async function one(ctx, next) {
  await next(); // <-- `two` will run at this point.

  if (ctx.body instanceof Error) {
    ctx.status = ctx.body.code || 500;
  }
});

app.use(async function two(ctx, next) {
  try {
    const result = await findSomeValue();
    ctx.body = result;
  } catch (err) {
    err.code = 404;
    ctx.body = err;
  }
});
```

In this contrived example, you can see an example of the status code being set by a middleware higher in the stack (`one`) based on the result of a middleware later in the stack (`two`). As a user I would expect this transaction to be named for `two`.